Examples/Io/Root/src/RootTrackParameterWriter.cpp

```diff
@@ -144,69 +189,140 @@ ActsExamples::ProcessCode ActsExamples::RootTrackParameterWriter::writeT(
-      const auto simHitIdx0 = indices.begin()->second;
-      const auto& simHit0 = *simHits.nth(simHitIdx0);
-      const auto p =
-          simHit0.momentum4Before().template segment<3>(Acts::eMom0).norm();
-      const auto& particleId = simHit0.particleId();
-      // The truth charge has to be retrieved from the sim particle
-      auto ip = particles.find(particleId);
-      if (ip != particles.end()) {
-        const auto& particle = *ip;
-        m_t_charge = static_cast<int>(particle.charge());
-        if (p != 0.0) {
-          m_t_qop = m_t_charge / p;
+      m_t_matched = true;
+      m_t_particleId = particleHitCounts.front().particleId.value();
+      m_nMajorityHits = particleHitCounts.front().hitCount;
+
+      // Get the index of the first space point
+      const auto& hitIdx = ptrack.front();
+      // Get the sim hits via the measurement to sim hits map
+      auto indices = makeRange(hitSimHitsMap.equal_range(hitIdx));
+      auto [truthLocal, truthPos4, truthUnitDir] =
+          averageSimHits(ctx.geoContext, surface, simHits, indices, logger());
+
+      // Get the truth track parameter at the first space point
+      m_t_loc0 = truthLocal[Acts::ePos0];
+      m_t_loc1 = truthLocal[Acts::ePos1];
+      m_t_phi = phi(truthUnitDir);
+      m_t_theta = theta(truthUnitDir);
+      m_t_qop = NaNfloat;
+      m_t_time = truthPos4[Acts::eTime];
+
+      m_t_charge = 0;
+      m_t_p = NaNfloat;
+      m_t_pt = NaNfloat;
+      m_t_eta = eta(truthUnitDir);
+
+      // momentum averaging makes even less sense than averaging position and
+      // direction. use the first momentum or set q/p to zero
+      if (!indices.empty()) {
+        // we assume that the indices are within valid ranges so we do not
+        // need to check their validity again.
+        const auto simHitIdx0 = indices.begin()->second;
+        const auto& simHit0 = *simHits.nth(simHitIdx0);
+        const auto p =
+            simHit0.momentum4Before().template segment<3>(Acts::eMom0);
+        const auto& particleId = simHit0.particleId();
+        // The truth charge has to be retrieved from the sim particle
+        if (auto ip = particles.find(particleId); ip != particles.end()) {
+          const auto& particle = *ip;
+          m_t_charge = static_cast<int>(particle.charge());
+          m_t_qop = particle.hypothesis().qOverP(p.norm(), particle.charge());
+          m_t_p = p.norm();
+          m_t_pt = perp(p);
         } else {
-          m_t_qop = std::numeric_limits<double>::quiet_NaN();
+          ACTS_WARNING("Truth particle with barcode " << particleId << "="
+                                                      << particleId.value()
+                                                      << " not found!");
         }
-      } else {
-        ACTS_DEBUG("Truth particle with barcode "
-                   << particleId << "=" << particleId.value() << " not found!");
       }
+
+      m_res_loc0 = m_loc0 - m_t_loc0;
```

Comment:
Couldn't we share the code for computing this with https://github.com/acts-project/acts/blob/aa536751435135cef7045dec4dd82b5f8cd5890d/Examples/Io/Root/src/RootTrackSummaryWriter.cpp#L414 ?

The computations are not terribly complicated, but I think it would be good to not introduce another place where this is computed... 

E.g., the periodic difference for angles is something that is missing here I think.

Reply:
Good point about the periodicity 👍 

I fear we have the residuals already in many more places than just here and in the `RootTrackSummaryWriter`. I personally don't think it is very bad to duplicate this because, like you say, it is quite trivial code but agree that it might be better to share it to avoid some pitfalls like the periodicity.

Since this happens in more places than two I would not make this part of this PR but of another potential follow up PR